### PR TITLE
llama : require first token to be BOS

### PR DESCRIPTION
**Pull request by ggerganov**
This is likely necessary to make the generation more accurate.
We first noticed this with running the new OpenLLaMA models. The generation completely fails if the first token is not `BOS`.

https://github.com/ggerganov/llama.cpp/issues/1291#issuecomment-1533260959

Setting the first token in each chunk of the perplexity computation to be `BOS` drives down the ppl values slightly (~0.05 for 7B), which indicates that  this is the right thing to do. Still, will be happy if somebody with better understanding chimes in and clarifies if we do need to enforce the first token to be `BOS`.

Another interesting observation is that the vanilla LLaMA models seem "resilient" to not having a `BOS`.
This seems to not be the case for OpenLLaMA. What is the difference that is causing this?

After merging this (or before), will recompute all perplexity values for 7B and 13B LLaMA.

Another effect from this change is that generation after context swap should be better, since before this change, we were "losing" the `BOS` token when `n_keep == 0` (i.e. default value).

---

### Perplexity after the change

| Model | Measure      | F16    | Q4_0   | Q4_1   | Q4_2   | Q5_0   | Q5_1   | Q8_0   |
|------:|--------------|-------:|-------:|-------:|-------:|-------:|-------:|-------:|
|    7B | perplexity   | 5.9066 | 6.1620 | 6.0910 | 6.1466 | 5.9862 | 5.9481 | 5.9069 |
|    7B | file size    |  13.0G |   4.0G |   4.8G |   4.0G |   4.4G |   4.8G |   7.1G |
|    7B | ms/tok @ 4th |    128 |     56 |     61 |     84 |     91 |     95 |     75 |
|    7B | ms/tok @ 8th |    128 |     47 |     55 |     48 |     53 |     59 |     75 |
|    7B | bits/weight  |   16.0 |    5.0 |    6.0 |    5.0 |    5.5 |    6.0 |    9.0 |
|   13B | perplexity   | 5.2543 | 5.3863 | 5.3607 | 5.3513 | 5.2856 | 5.2706 | 5.2548 |
|   13B | file size    |  25.0G |   7.6G |   9.1G |   7.6G |   8.4G |   9.1G |    14G |
|   13B | ms/tok @ 4th |    239 |    104 |    113 |    160 |    176 |    185 |    141 |
|   13B | ms/tok @ 8th |    240 |     85 |     99 |     97 |    108 |    117 |    147 |
|   13B | bits/weight  |   16.0 |    5.0 |    6.0 |    5.0 |    5.5 |    6.0 |    9.0 |

#### For reference - before the change

| Model | Measure      | F16    | Q4_0   | Q4_1   | Q4_2   | Q5_0   | Q5_1   | Q8_0   |
|------:|--------------|-------:|-------:|-------:|-------:|-------:|-------:|-------:|
|    7B | perplexity   | 5.9565 | 6.2103 | 6.1286 | 6.1698 | 6.0139 | 5.9934 | 5.9571 |
|    7B | file size    |  13.0G |   4.0G |   4.8G |   4.0G |   4.4G |   4.8G |   7.1G |
|    7B | ms/tok @ 4th |    128 |     56 |     61 |     84 |     91 |     95 |     75 |
|    7B | ms/tok @ 8th |    128 |     47 |     55 |     48 |     53 |     59 |     75 |
|    7B | bits/weight  |   16.0 |    5.0 |    6.0 |    5.0 |    5.5 |    6.0 |    9.0 |
|   13B | perplexity   | 5.2455 | 5.3748 | 5.3471 | 5.3433 | 5.2768 | 5.2582 | 5.2458 |
|   13B | file size    |  25.0G |   7.6G |   9.1G |   7.6G |   8.4G |   9.1G |    14G |
|   13B | ms/tok @ 4th |    239 |    104 |    113 |    160 |    176 |    185 |    141 |
|   13B | ms/tok @ 8th |    240 |     85 |     99 |     97 |    108 |    117 |    147 |
|   13B | bits/weight  |   16.0 |    5.0 |    6.0 |    5.0 |    5.5 |    6.0 |    9.0 |

### CODE REVIEW

**Comment by young-geng** [comment]
Hi! I'm one of the creators of OpenLLaMA. We have realized that our model is indeed sensitive to BOS token and many existing implementations do not prepend the BOS token at generation time. To make our model more compatible with existing implementations, we have released our [new 300b token checkpoint that is less sensitive to BOS tokens](https://huggingface.co/openlm-research/open_llama_7b_preview_300bt).

**Review by glinscott — APPROVED** [review]

**Comment by ggerganov** [comment]
@young-geng Thanks for the info

@glinscott Good point - will fix this

The new perplexity results are out:

- for 7B the ppl goes down ~`-0.04` across all quantization modes
- for the 13B, the ppl increases slightly by ~`0.01`

I think the decision of whether to always have a `BOS` token at the start of the context depends on how the training was done. If all sequences during training had a `BOS` token, then this is what the model has "seen" and we have to always make sure it is there. If not, then we can keep the original implementation as it is on `master`.

My guess is that during training, all sequences do have a `BOS` token at the start - just need someone to confirm this

**Comment by young-geng** [comment]
For OpenLLaMA, we always add BOS to the sequence during training. We believe that they also did this for the official LLaMA, as adding BOS improves a few percent on NLP benchmarks.

**Comment by klosax** [comment]
open llama 300bt q5_1 perplexity on wiki.test.raw (ctx 512 n_batch 512)

without this pr: 11.31323402
with this pr: 11.25677669

**Comment by ggerganov** [comment]
I think this should handle smaller batch sizes correctly now

**Comment by klosax** [comment]
> I think this should handle smaller batch sizes correctly now

I tested this and the save/restore of the original token does not seem to make any difference in ppl.
On LLaMA 7B q5_1 using ctx 512 / batch 512 the ppl is the same with or without saving the token: 5.9479

**Comment by glinscott** [comment]
Looks good, thanks!  And interesting that it doesn't have any impact, @klosax did you test with eg. batch size 8?  that should have had a significant impact with the previous implementation.

**Review by glinscott — APPROVED** [review]

**Comment by klosax** [comment]
> Looks good, thanks! And interesting that it doesn't have any impact, @klosax did you test with eg. batch size 8? that should have had a significant impact with the previous implementation.

Yes indeed, the BOS token should only be added to the first batch of each ctx chunk, not to every batch. Saving/restoring the original token does not make any difference to the ppl, as the softmax does not use the first half of the ctx chunk tokens.

My testing also indicate that the ppl goes down even further if the BOS token is properly inserted instead of simply overwriting the original tokens.

**Comment by ggerganov** [comment]
The initial implementation in this branch indeed was broken for smaller batches because I was putting the `BOS` token for each batch of the chunk instead of only just in the first batch. After the fix in [7f33230](https://github.com/ggerganov/llama.cpp/pull/1303/commits/7f33230a4088bf9c2f1d5e60dfce1744a84eae56) it should be good, as reported by @klosax
